### PR TITLE
[6.x] Add loadFactoriesFrom method

### PR DESCRIPTION
**Pull request by lexdewilligen**
This PR introduces a `loadFactoriesFrom` method which can be used by Package developers to support development with Models or Migrations that are provided by the package. We've already got several supporting methods like `loadRoutesFrom`, `loadTranslationsFrom` and `loadMigrationsFrom`. I feel this would also be a good functionality to add to that list.

Let's say a package provides you a User model along with a user migration. This method would allow package developers to also provide a factory for this model from within their package. Developers using this package could easily seed the Users table or test their application without having to write a factory themselves.

### CODE REVIEW

**Comment by crynobone** [comment]
> Let's say a package provides you a User model along with a user migration. This method would allow package developers to also provide a factory for this model from within their package.

If you are using `orchestra/testbench` for your package development testing we already natively provide support for loading factories.

**Comment by gocanto** [comment]
This has been achievable since long time ago, see this example for an work around
https://github.com/ant-vel/App/blob/master/database/factories/AntvelFactory.php#L12

**Comment by crynobone** [comment]
> Developers using this package could easily seed the Users table or test their application without having to write a factory themselves.

Developers using `laravel/laravel` should already have a `UserFactory`, having packages trying to create an alternative `UserFactory` for developers may produce inconsistency, what happen when 5 packages trying to setup `UserFactory`?

**Comment by lexdewilligen** [comment]
> > Developers using this package could easily seed the Users table or test their application without having to write a factory themselves.
> 
> Developers using `laravel/laravel` should already have a `UserFactory`, having packages trying to create an alternative `UserFactory` for developers may produce inconsistency, what happen when 5 packages trying to setup `UserFactory`?

It was just an example where they provide a factory for a model within their package so that developers don't have to write that themselves.

File: src/Illuminate/Support/ServiceProvider.php [hunk 1]
```diff
@@ -136,6 +137,18 @@ protected function loadMigrationsFrom($paths)
         });
     }
 
+    /**
+     * Register factories.
+     *
+     * @param string $path
+     * @return void
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
```
Comment (driesvints) [inline src/Illuminate/Support/ServiceProvider.php]: Remove this line. This method doesn't throws the exception.

File: src/Illuminate/Support/ServiceProvider.php [hunk 2]
```diff
@@ -136,6 +137,18 @@ protected function loadMigrationsFrom($paths)
         });
     }
 
+    /**
+     * Register factories.
+     *
+     * @param string $path
```
Comment (driesvints) [inline src/Illuminate/Support/ServiceProvider.php]:
```suggestion
     * @param  string  $path
```

**Review by driesvints — DISMISSED** [review]

**Comment by lexdewilligen** [comment]
> > Let's say a package provides you a User model along with a user migration. This method would allow package developers to also provide a factory for this model from within their package.
> 
> If you are using `orchestra/testbench` for your package development testing we already natively provide support for loading factories.

Yes, however, some developers like to use factories for seeding their database when developing, so therefore maybe it's a good addition to include in Laravel itself. Furthermore, some developers may not use `orchestra/testbench` for some reason.

**Comment by lexdewilligen** [comment]
> Nice work.
> 
> This has been achievable since long time ago, see this example for an work around
> https://github.com/ant-vel/App/blob/master/database/factories/AntvelFactory.php#L12
> 
> Plus this indeed
> 
> > If you are using orchestra/testbench for your package development testing we already natively provide support for loading factories.

Well yes there are several workarounds but all with downsides. In your suggested workaround, developers still have to create a factory to load the package factory, which is extra files and work. I love packages where you just install it and don't have to do anything since migrations, factories, translations, views etc. are all loaded automatically.

**Comment by driesvints** [comment]
@lexdewilligen I fixed the build on the 6.x branch. Can you rebase? The build will pass then.

**Comment by driesvints** [comment]
@lexdewilligen you're using merge commits instead of rebasing so the commit history looks a bit messy. Can you try to rebase instead?

**Comment by crynobone** [comment]
>I love packages where you just install it and don't have to do anything since migrations, factories, translations, views etc. are all loaded automatically.

Migrations, Translations and views is required in the production environment, while factories are meant for development (testing and local prototyping) purposes.

**Comment by lexdewilligen** [comment]
I didn't know how to do so, should I revert the merges now?

**Comment by driesvints** [comment]
@lexdewilligen it's okay to leave them in if you don't know how to rebase.

Here's a tutorial on rebasing in git: https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

**Comment by driesvints** [comment]
I also have to say I agree with @crynobone on this one.

**Comment by lexdewilligen** [comment]
> However for this proposal, Illuminate\Database\Eloquent\Factory::load() will straightaway require all the factories files regardless whether you need it or not.

@crynobone I agree, good point, can't think of a good argument to beat you there! However, I still believe this to be a handy feature.

**Comment by crynobone** [comment]
```php
if ($this->app->runningInConsole()) {
    $this->publishes([
        __DIR__.'/../factories/FooFactory.php' => database_path('factories/FooFactory.php'),
    ], 'factories');
 }
```

Developer has full control of the factory, if they want to make any modification (alter the migrations etc) they can modify the factory without editing the `vendor/` files.

**Comment by lexdewilligen** [comment]
> ```
> if ($this->app->runningInConsole()) {
>     $this->publishes([
>         __DIR__.'/../factories/FooFactory.php' => database_path('factories/FooFactory.php'),
>     ], 'factories');
>  }
> ```
> 
> Developer has full control of the factory, if they want to make any modification (alter the migrations etc) they can modify the factory without editing the `vendor/` files.

Yes but then you would have to publish the factory files which is just the thing I'm trying to prevent. I'd like to use the factory without having to write or publish it in my own factories folder.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Pull request closed without merging**